packages/core/src/signals/src/weak_ref.ts

```diff
@@ -0,0 +1,23 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {global} from '../../util/global';
+
+// `WeakRef` is not always defined in every TS environment where Angular is compiled. Instead,
+// alias it as a local export by reading it off of the global context.
+
+export interface WeakRef<T extends object> {
+  deref(): T|undefined;
+}
+
+export interface WeakRefCtor {
+  new<T extends object>(value: T): WeakRef<T>;
+}
+
+// tslint:disable-next-line: no-toplevel-property-access
+export const WeakRef: WeakRefCtor = global['WeakRef'];
```

Comment:
This is perhaps a bad idea, but rather than eagerly creating a `WeakRef` on each `Producer`/`Consumer`, you could use a single shared cache here to be able to lazily create the `WeakRef` while still not affecting the monomorphism of the `Producer`/`Consumer`
```suggestion
const WeakRef: WeakRefCtor = global['WeakRef'];

const weakRefCache = new WeakMap<object, WeakRef<object>>();
export function CachedWeakRef<T extends object>(value: T): WeakRef<T> {
    return (weakRefCache.get(value) ?? weakRefCache.set(value, new WeakRef(value)).get(value)) as WeakRef<T>;
}
```

Reply:
This was indeed a bad idea.
I ran some benchmarks on this using one of our real dependency graphs, and it was 25% slower, and cost 5% more memory.

Reply:
Thnx for checking this @DavidANeil